Implement grouping of the stack frames

gzsombor opened the pull request:
The rules how the frames are categorized are not editable as of now, just a single action is visible to switch on-off the behaviour

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See https://github.com/eclipse-jdt/.github/security/policy
-->

## What it does
 Groups - and hides - the supposedly not relevant stack frames from the debug views.  
![before](https://github.com/user-attachments/assets/54e0c830-1d12-4c90-8eaf-b9339a44c192)
![Screenshot_20250321_092524](https://github.com/user-attachments/assets/df56dc75-754d-4b72-9f9d-2543aec1acf8)

After 

![new_group1](https://github.com/user-attachments/assets/00334ab0-681e-42ae-b916-b332e2578aed)
Expanding the group:
![new_group2](https://github.com/user-attachments/assets/f7cb0578-59d9-4e6b-abfb-8cc6690d5ca4)


## How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Start debugging, and notice that instead of having hundreds of frames only a couple are visible

## Author checklist

- [X] I have thoroughly tested my changes
- [X] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [X] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

gzsombor commented:
@jukzi : Could you please have a look on this when you have time and check what needs to be done to be merged?

jukzi commented:
unfortunately i am very limited in time. @SougandhS can you test/review this PR?

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/JDIModelPresentation.java:
```diff
@@ -936,6 +947,25 @@ protected Image getDebugElementImage(Object element) {
 	}
 
+	private Image getStackFrameImage(IJavaStackFrame stackFrame) {
+		var image = getStackFrameProvider().getStackFrameImage(stackFrame);
```
please include java doc for `getStackFrameImage`

gzsombor replied:
done

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/JDIModelPresentation.java:
```diff
@@ -2048,6 +2069,12 @@ protected JavaElementLabelProvider getJavaLabelProvider() {
 	}
 
+	private StackFramePresentationProvider getStackFrameProvider() {
+		if (fStackFrameProvider == null) {
```
missing java doc

gzsombor replied:
done

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java:
```diff
@@ -0,0 +1,193 @@
+/*******************************************************************************
+ * Copyright (c) 2021 Zsombor Gegesy.
+ *
```
copyright year is wrong.
please include " and others" after your name

gzsombor replied:
Believe it or not, I wrote that class in 2021 :smiling_face_with_tear: - but I've adjusted to include 2025 too as well.

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java:
```diff
@@ -0,0 +1,193 @@
+		collapseStackFrames = store.getBoolean(IJDIPreferencesConstants.PREF_COLLAPSE_STACK_FRAMES);
+	}
+
+	private Filters getActivePlatformFilters(IPreferenceStore store) {
```
Missing java doc

gzsombor replied:
done

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java:
```diff
@@ -0,0 +1,193 @@
+		return new Filters(PLATFORM_STACK_FRAMES.getActiveList(store));
+	}
+
+	private Filters getActiveCustomFilters(IPreferenceStore store) {
```
missing java doc

gzsombor replied:
done

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java:
```diff
@@ -0,0 +1,193 @@
+	/**
+	 * Visible for testing!
+	 */
+	public IJavaStackFrame.Category categorize(IJavaStackFrame frame) throws DebugException {
```
please include proper java doc

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java:
```diff
@@ -0,0 +1,193 @@
+	}
+
+	private Category categorizeSourceElement(IJavaStackFrame frame) {
+		var sourceLocator = frame.getLaunch().getSourceLocator();
```
Missing java doc

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java:
```diff
@@ -0,0 +1,193 @@
+		}
+	}
+
+	public boolean isCollapseStackFrames() {
```
missing java doc

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/actions/CollapseStackFramesAction.java:
```diff
@@ -0,0 +1,34 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Zsombor Gegesy and others.
```
please update copyright year

gzsombor replied:
done

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/monitors/JavaThreadContentProvider.java:
```diff
@@ -59,18 +71,32 @@ protected int getChildCount(Object element, IPresentationContext context, IViewe
 	}
 
+	private int getFrameCount(IJavaThread thread) throws DebugException {
+		if (getStackFrameProvider().isCollapseStackFrames()) {
```
Missing java doc

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/monitors/JavaThreadContentProvider.java:
```diff
@@ -59,18 +71,32 @@ protected int getChildCount(Object element, IPresentationContext context, IViewe
 	}
 
+	private Object[] getChildren(GroupedStackFrame parent, int index, int length) {
+		return parent.getFramesAsArray(index, length);
```
missing java doc

gzsombor replied:
done

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/monitors/JavaThreadContentProvider.java:
```diff
@@ -81,40 +107,65 @@ protected Object[] getChildren(IJavaThread thread) {
 	}
 
+	private List<Object> getStackFrames(IJavaThread thread) throws DebugException {
+		var frames = thread.getStackFrames();
```
please include java doc

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/monitors/JavaThreadContentProvider.java:
```diff
@@ -174,5 +228,11 @@ private ISchedulingRule getThreadRule(IViewerUpdate[] updates) {
 	}
 
+	private synchronized StackFramePresentationProvider getStackFrameProvider() {
+		if (stackFrameProvider == null) {
```
missing java doc

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/monitors/MonitorsAdapterFactory.java:
```diff
@@ -40,7 +41,7 @@ public class MonitorsAdapterFactory implements IAdapterFactory {
 
     	if (IElementContentProvider.class.equals(adapterType)) {
-    		if (adaptableObject instanceof IJavaThread) {
+			if (adaptableObject instanceof IJavaThread || adaptableObject instanceof GroupedStackFrame) {
```
Some additional space was added here

gzsombor replied:
Yeah, sorry. Files with mixed tabs and spaces are very confusing and it is easy to mess up them even more!

SougandhS commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/GroupedStackFrame.java:
```diff
@@ -0,0 +1,60 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Zsombor Gegesy.
+ *
```
please include " and others"

gzsombor replied:
done.

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/JDIDebugUIPreferenceInitializer.java:
```diff
@@ -46,6 +46,12 @@ public void initializeDefaultPreferences() {
+		store.setDefault(IJDIPreferencesConstants.PREF_ACTIVE_CUSTOM_FRAME_FILTER_LIST, ""); //$NON-NLS-1$
+		store.setDefault(IJDIPreferencesConstants.PREF_INACTIVE_CUSTOM_FRAME_FILTER_LIST, ""); //$NON-NLS-1$
+		store.setDefault(IJDIPreferencesConstants.PREF_COLLAPSE_STACK_FRAMES, true);
+
```
Try using `org.eclipse.jface.util.Util.ZERO_LENGTH_STRING` for empty string declarations, so you can avoid `NON-NLS` comments

gzsombor replied:
Thanks, I didn't know about this class!

SougandhS commented:
Stack icons looks bit odd

<img width="120" alt="image" src="https://github.com/user-attachments/assets/736efff1-4049-443f-aa44-315b4f21b2d0" />

Its same as  .Class files
<img width="191" alt="image" src="https://github.com/user-attachments/assets/7dbe92af-146b-4f6c-8ece-5e5f4523e7fc" />

On master - 
<img width="91" alt="image" src="https://github.com/user-attachments/assets/7200dbae-4d23-4b18-9188-9ce58d0b5629" />

SougandhS commented:
if the breakpoint is in .class file then original stack (master) icons are shown

<img width="455" alt="image" src="https://github.com/user-attachments/assets/7df90703-5383-4b5a-ada6-cf33108961e8" />

gzsombor commented:
> if the breakpoint is in .class file then original stack (master) icons are shown
> <img alt="image" width="455" src="https://private-user-images.githubusercontent.com/177713286/402901165-7df90703-5383-4b5a-ada6-cf33108961e8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY4OTQzMDMsIm5iZiI6MTczNjg5NDAwMywicGF0aCI6Ii8xNzc3MTMyODYvNDAyOTAxMTY1LTdkZjkwNzAzLTUzODMtNGI1YS1hZGE2LWNmMzMxMDg5NjFlOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExNFQyMjMzMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMzg0MDEyMjczYmQ3N2M0ODFjZGE0NjI4ZTk2ZmM2ZGUxZGM5MjU3ZTk0NGQ4NjA4OTdiNzFlODkxMjJkMzk5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pTb1Ca1-PiM4vY0m5iTJLTcw_nmSzewpdxMzXYRFGoE">
> 
> both <img alt="image" width="264" src="https://private-user-images.githubusercontent.com/177713286/402903811-0ee6b98f-a9f1-43fc-b861-4577bff55362.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY4OTQzMDMsIm5iZiI6MTczNjg5NDAwMywicGF0aCI6Ii8xNzc3MTMyODYvNDAyOTAzODExLTBlZTZiOThmLWE5ZjEtNDNmYy1iODYxLTQ1NzdiZmY1NTM2Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDExNFQyMjMzMjNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMzhlMGZkYzljY2EwODJjYjEwMjczMDI4Y2NmMGY2OTk2YjQ4ZDk0ODMzNWNmOGRkZjAyMzVkNzAwNDYxMmYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4rsyMUkolwkEwYvbm9Pk9U_WiDOBEYpI9JwE--0kyXk">
> 
> Feature looks awsm! Would be nice if the icons are consistent

I'm happy that you liked the feature!
 It is a long time, since I've written most of the code - I even forgot, that I've modified that part which returns images for the stack frames. I've modified again a bit, so the icon will be the same old, for all the prod/test codes, and have the 'jar' icon for classes coming from the libraries and 'platform'.

SougandhS commented:
>have the 'jar' icon for classes coming from the libraries and 'platform'.

<img width="342" alt="image" src="https://github.com/user-attachments/assets/4780f31f-bb7f-4272-867d-b1c8a858ae93" />

Icons looks good now thanks!

SougandhS commented on org.eclipse.jdt.debug.ui/plugin.xml:
```diff
@@ -2208,6 +2208,15 @@
+        style="toggle"
+        menubarPath="org.eclipse.jdt.debug.ui.LaunchView.javaSubmenu/javaPart"
+        id="org.eclipse.jdt.debug.ui.launchViewActions.CollapseStackFrames"/>
+
```
additional space you can remove

SougandhS commented:
Hi @gzsombor , could pls take the latest pull and squash everything to one commit only

SougandhS reviewed (approved):
LGTM 👍

jukzi commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java:
```diff
@@ -0,0 +1,214 @@
+		} catch (DebugException e) {
+			JDIDebugUIPlugin.log(e);
+		}
+		frame.setCategory(result);
```
why does a getter getCategory **set** something? that is unexpected

jukzi replied:
so if setCategory is only used as cache please move the caching implementation into the getter to avoid the setter. However it feels like the implementation is wrong in means of it will not update the cached value if the options change.

gzsombor replied:
The setter was there, because all the settings were managed by JDIDebugUIPlugin, not by JDIDebugPlugin, partially because managing class/package filter lists are a bit easier with org.eclipse.jface.preference.IPreferenceStore and org.eclipse.jdt.internal.ui.filtertable.FilterManager - those are UI classes, not reachable from the backend.
 But I could rewrote it, the base functionality seems to be working - I still have to refactor the settings page and the couple of tests, but that will come in the other PR.

jukzi commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java:
```diff
@@ -0,0 +1,214 @@
+	/**
+	 * Class to decide if a particular class name is part of a list of classes and list of packages.
+	 */
+	static class Filters {
```
looks like a record

gzsombor replied:
That's true! When I first wrote this class, Java 16 was not allowed in JDT.

jukzi commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/debug/core/IJavaStackFrame.java:
```diff
@@ -533,4 +570,18 @@ public IJavaVariable findVariable(String variableName)
+	 * @param category
+	 *            the new category of the stack frame.
+	 */
+	public void setCategory(Category category);
```
can it somehow avoided to have such setter? it is not obvious to me if such setter should `fireChangeEvent(DebugEvent.STATE);`

gzsombor replied:
Somehow we need to store the information how that frame was categorized - even if the process is not slow, it needs to lookup the source, and check which classpath entry contains that, etc.
In theory, we could have a map from IJavaStackFrame -> Category in some singleton, but it's unclear how we can detect, if that stack frame was not accessible anymore, so maybe a WeakHashMap could work, but it felt too complicated and error prone.
 This setter is only called once - when we first want to display the frame, so firing a change event should not needed - at least it seems to work fine without it.

jukzi commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/StackFramePresentationProvider.java:
```diff
@@ -0,0 +1,214 @@
+		}
+	}
+
+	private Filters platform;
```
can those be declared final or even use a record?

gzsombor replied:
The variables inside StackFramePresentationProvider are needs to be mutable, but Filters can be turned into a record.

jukzi commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/monitors/JavaThreadContentProvider.java:
```diff
@@ -81,40 +107,71 @@ protected Object[] getChildren(IJavaThread thread) {
 				}
 			}
-			IStackFrame[] frames = thread.getStackFrames();
+			var frames = getStackFrames(thread);
```
I have a weak opinion about using "var" is hard to review the code.

jukzi commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/debug/core/IJavaStackFrame.java:
```diff
@@ -35,6 +35,43 @@
+	 * @since 3.22
+	 *
+	 */
+	enum Category {
```
It feels wrong to publish this enum as API as it may be subject to change. Especially "platform" may be usefull but has no clear definition. Also those categories may not be disjunct. code could be synthetic platform test and library at once. One could either use plain Strings as Category or have a CategoryRegistry where new Categories could be added at runtime.

gzsombor replied:
Yes, the categories are not exact and there could be frames that could belong to multiple categories.
However, I don't think that adding new categories is a feasible - and useful feature ever. Main problem is that a single category needs to be handled in multiple places and going for a very general solution would be hard. For example, to have a foreground and background colors we have to declare them in the plugin.xml, otherwise it can't be set in the color editing preference page.
Maybe, we could have Category as a 'record' and have them declared as a fixed constant in the JDIDebugPlugin.

jukzi replied:
As long as the solution does not make the categories public API, it may be ok.

gzsombor replied:
What do you think about this solution?

jukzi commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIStackFrame.java:
```diff
@@ -1730,4 +1732,12 @@ public void forceReturn(IJavaValue value) throws DebugException {
+	@Override
+	public Category getCategory() {
+		if (fCategory == null) {
+			fCategory = JDIDebugPlugin.getDefault().getStackFrameCategorizer().categorize(this);
```
that's a better cache mechanism :-) as i guess this code is called from different threads it needs some ynchronization

gzsombor replied:
synchronization added

gzsombor commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/monitors/JavaThreadContentProvider.java:
```diff
@@ -81,40 +107,71 @@ protected Object[] getChildren(IJavaThread thread) {
+	 * expanded on demand by the user.
+	 */
+	private List<Object> getStackFrames(IJavaThread thread) throws DebugException {
+		var frames = thread.getStackFrames();
```
I feel mostly they just add to the noise that hinders understanding.
I've removed where you mentioned.

SougandhS commented:
Hi @gzsombor, pls squash all commits into one

gzsombor commented:
> Hi @gzsombor, pls squash all commits into one

Done and also the build is green now!

jukzi commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/debug/core/JDIDebugModel.java:
```diff
@@ -149,6 +152,61 @@ public class JDIDebugModel {
+	 *
+	 * @since 3.22
+	 */
+	public static final Category CATEGORY_CUSTOM_FILTERED = new Category("CUSTOM_FILTERED", false); //$NON-NLS-1$
```
As said i won't accept any particular Category as API. Can you make the List of Categories somehow non-api?

gzsombor replied:
I've removed the list of Categories. I don't think removing the String/Category constants would make sense, but if you prefer, I can move those constants from JDIDebugModel to StackFrameCategorizer for example

jukzi replied:
When we would make something API it would stay almost forever. That should only be done for things that are so well crafted that they are unlikely to change. This Categories are not set to stone.

gzsombor replied:
What makes something into an API in your opinion? As the logic of deciding of the categories is in a different module from the code which needs those categories to actually pick the different colors, or to show the UI to configure these categories, in the end, we need to have some public constants that ensures that they are referring to the same thing.
 What do you suggest what to do?

jukzi replied:
A API is a thing that can be accessed from another plugin. For example a public class/member in a public package.
Sorry, I don't have the time to think about a concrete solution for this PR. I can just guess, that it would be enough if either the UI or the core knows all categories at compile time and tells the other plugin at runtime what it wants.

gzsombor commented:
It is all green now, could you please @jukzi have another look on it?

SougandhS commented:
> It is all green now, could you please @jukzi have another look on it?

He has moved out of the Eclipse team

SougandhS commented:
please update the branch

gzsombor commented:
> please update the branch

I've rebased it again.

gzsombor commented:
Who can review and merge this PR?

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/JDIDebugUIPreferenceInitializer.java:
```diff
@@ -46,6 +46,14 @@ public void initializeDefaultPreferences() {
 		store.setDefault(IJDIPreferencesConstants.PREF_STEP_THRU_FILTERS, true);
 
+		// Grouping stack frames
+		store.setDefault(IJDIPreferencesConstants.PREF_COLLAPSE_STACK_FRAMES, true);
```
Could you please put set default as `false`

SougandhS reviewed (changes requested):
Please provide an icon for context menu <br> 
<img width="453" alt="image" src="https://github.com/user-attachments/assets/8873c29b-5cdf-41e3-aff4-150f65375d1c" />

SougandhS reviewed (approved):
Please update the context menu screenshot in PR description

gzsombor commented:
I've added a screenshot about the menu item.

SougandhS commented:
Thank you for contributing @gzsombor , its a really useful feature. 👍
